plugin/pkg/admission/initialresources/admission.go

```diff
@@ -0,0 +1,138 @@
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package initialresources
+
+import (
+	"flag"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/admission"
+	"k8s.io/kubernetes/pkg/api"
+	apierrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/resource"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+)
+
+var (
+	source     = flag.String("ir_data_source", "influxdb", "")
+	percentile = flag.Int64("ir_percentile", 90, "")
+)
+
+const (
+	week  = 7 * 24 * time.Hour
+	month = 30 * 24 * time.Hour
+)
+
+func init() {
+	admission.RegisterPlugin("InitialResources", func(client client.Interface, config io.Reader) (admission.Interface, error) {
+		s, err := newDataSource(*source)
+		if err != nil {
+			return nil, err
+		}
+		return newInitialResources(s), nil
+	})
+}
+
+type initialResources struct {
+	*admission.Handler
+	source dataSource
+}
+
+func newInitialResources(source dataSource) admission.Interface {
+	return &initialResources{
+		Handler: admission.NewHandler(admission.Create),
+		source:  source,
+	}
+}
+
+func (ir initialResources) Admit(a admission.Attributes) (err error) {
+	// Ignore all calls to subresources or resources other than pods.
+	if a.GetSubresource() != "" || a.GetResource() != string(api.ResourcePods) {
+		return nil
+	}
+	pod, ok := a.GetObject().(*api.Pod)
+	if !ok {
+		return apierrors.NewBadRequest("Resource was marked with kind Pod but was unable to be converted")
+	}
+
+	ir.estimateResourcesIfNotSet(pod)
+	return nil
+}
+
+func (ir initialResources) estimateResourcesIfNotSet(pod *api.Pod) {
+	for i := range pod.Spec.Containers {
+		c := &pod.Spec.Containers[i]
+		req := c.Resources.Requests
+		lim := c.Resources.Limits
+		var cpu, mem *resource.Quantity
+		if _, ok := req[api.ResourceCPU]; !ok {
+			if _, ok2 := lim[api.ResourceCPU]; !ok2 {
+				cpu = ir.getEstimation(api.ResourceCPU, c)
+			}
+		}
+		if _, ok := req[api.ResourceMemory]; !ok {
+			if _, ok2 := lim[api.ResourceMemory]; !ok2 {
+				mem = ir.getEstimation(api.ResourceMemory, c)
+			}
+		}
+
+		// If Requests doesn't exits and an estimation was made, create Requests.
+		if req == nil && (cpu != nil || mem != nil) {
+			c.Resources.Requests = api.ResourceList{}
+			req = c.Resources.Requests
+		}
+		if cpu != nil {
+			glog.Infof("CPU estimation for container %v in pod %v/%v is %v", c.Name, pod.ObjectMeta.Namespace, pod.ObjectMeta.Name, cpu.String())
+			req[api.ResourceCPU] = *cpu
+		}
+		if mem != nil {
+			glog.Infof("Memory estimation for container %v in pod  %v/%v is %v", c.Name, pod.ObjectMeta.Namespace, pod.ObjectMeta.Name, mem.String())
+			req[api.ResourceMemory] = *mem
+		}
+	}
+	// TODO: verify the estimates fits in LimitRanger
```

Comment:
If the InitialResources set mem to 1GB and the max allowed amount is 512MB the pod will be reject. To prevent users from such unexpected behavior I'm planning to set prediction as min(max(minAllowed, estimation), maxAllowed). This TODO is about it.